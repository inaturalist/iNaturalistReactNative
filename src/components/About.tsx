import { useNavigation } from "@react-navigation/native";
import {
  Body1,
  Body2,
  Button,
  Heading4,
  ScrollViewWrapper,
  Tabs,
  UnderlinedLink,
  ViewWrapper
} from "components/SharedComponents";
import { Image, Pressable, View } from "components/styledComponents";
import { t } from "i18next";
import React, { useState } from "react";
import { getBuildNumber, getVersion } from "react-native-device-info";
import { useDebugMode } from "sharedHooks";

const aboutID = "about";
const teamID = "team";

const About = ( ) => {
  const navigation = useNavigation();
  const [activeTab, setActiveTab] = useState( aboutID );
  const [count, setCount] = useState( 0 );
  const appVersion = getVersion();
  const buildVersion = getBuildNumber();
  const { isDebug, toggleDebug } = useDebugMode( );

  const onTermsPressed = async () => {
    const url = "https://www.inaturalist.org/pages/terms";
    navigation.navigate( "FullPageWebView", {
      title: t( "TERMS-OF-USE" ),
      initialUrl: url,
      loggedIn: false
    } );
  };

  const onPrivacyPressed = async () => {
    const url = "https://www.inaturalist.org/pages/privacy";
    navigation.navigate( "FullPageWebView", {
      title: t( "PRIVACY-POLICY" ),
      initialUrl: url,
      loggedIn: false
    } );
  };

  const onCommunityPressed = async () => {
    const url = "https://www.inaturalist.org/pages/community+guidelines";
    navigation.navigate( "FullPageWebView", {
      title: t( "COMMUNITY-GUIDELINES" ),
      initialUrl: url,
      loggedIn: false
    } );
  };

  const onVersionPressed = () => {
    if ( ( count + 1 ) % 3 === 0 ) {
      toggleDebug();
    }
    setCount( count + 1 );
  };

  return (
    <ViewWrapper>
      <Tabs
        tabs={[
          {
            id: aboutID,
            text: t( "ABOUT" ),
            onPress: () => {
              setActiveTab( aboutID );
            }
          },
          {
            id: teamID,
            text: t( "TEAM" ),
            onPress: () => {
              setActiveTab( teamID );
            }
          }
        ]}
        activeId={activeTab}
      />
      <ScrollViewWrapper>
        {activeTab === aboutID && (
          <View className="px-4 py-8">
            <Image
              className="w-[65%] max-h-12 xl:max-h-24 2xl:max-h-32 object-center self-center mb-8"
              resizeMode="contain"
              source={require( "images/inaturalist-dark.png" )}
              accessibilityIgnoresInvertColors
            />
            <Heading4 className="mb-3">
              {t( "INATURALIST-MISSION-VISION" )}
            </Heading4>
            <Body2 className="mb-5">
              {t( "iNaturalist-mission-is-to-connect" )}
            </Body2>
            <Body2 className="mb-8">{t( "iNaturalists-vision-is-a-world" )}</Body2>
            <Heading4 className="mb-3">{t( "WHAT-IS-INATURALIST" )}</Heading4>
            <Body2 className="mb-5">
              {t( "iNaturalist-helps-you-identify" )}
            </Body2>
            <Body2 className="mb-5">{t( "Whats-more-by-recording" )}</Body2>
            <Body2 className="mb-8">{t( "iNaturalist-is-supported-by" )}</Body2>
            <UnderlinedLink
              className="mb-5"
              onPress={() => onTermsPressed()}
            >
              {t( "Terms-of-Use" )}
            </UnderlinedLink>
            <UnderlinedLink
              className="mb-5"
              onPress={() => onPrivacyPressed()}
            >
              {t( "Privacy-Policy" )}
            </UnderlinedLink>
            <UnderlinedLink
              className="mb-8"
              onPress={() => onCommunityPressed()}
            >
              {t( "Community-Guidelines" )}
            </UnderlinedLink>
            <Pressable
              accessibilityRole="button"
              className="items-center justify-center"
              onPress={() => onVersionPressed()}
            >
              <Body1>{ t( "Version-app-build", { appVersion, buildVersion } )}</Body1>
            </Pressable>
            {isDebug && (
              <Button
                text="TURN OFF DEBUG MODE"
                className="mt-5 bg-deeppink"
                level="primary"
                onPress={() => {
                  setCount( 0 );
                  toggleDebug();
                }}
              />
            )}
          </View>
        )}
        {activeTab === teamID && (
          <View className="px-4 py-8">
            <Heading4 className="mb-3">{t( "INATURALIST-TEAM" )}</Heading4>
            <Body2 className="mb-8">
              {t( "iNaturalists-apps-are-designed-and-developed-3" )}
            </Body2>
            <Heading4 className="mb-3">{t( "INATURALIST-COMMUNITY" )}</Heading4>
            <Body2 className="mb-8">
              {t( "iNaturalist-is-supported-by-our-community" )}
            </Body2>
            <Heading4 className="mb-3">{t( "INATURALIST-NETWORK" )}</Heading4>
            <Body2 className="mb-8">{t( "The-iNaturalist-Network" )}</Body2>
            <Heading4 className="mb-3">{t( "COLLABORATORS" )}</Heading4>
            <Body2>{t( "The-models-that-suggest-species" )}</Body2>
          </View>
        )}
      </ScrollViewWrapper>
    </ViewWrapper>
  );
};

export default About;
