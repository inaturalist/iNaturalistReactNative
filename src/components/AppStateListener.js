import { focusManager } from "@tanstack/react-query";
import useDeviceStorageFull from "components/Camera/hooks/useDeviceStorageFull";
import { useEffect, useState } from "react";
import { AppState } from "react-native";
import { log } from "sharedHelpers/logger";
import {
  usePerformance
} from "sharedHooks";
import { isDebugMode } from "sharedHooks/useDebugMode";

const logger = log.extend( "AppStateListener" );

const AppStateListener = ( ) => {
  const { loadTime } = usePerformance( {
    screenName: "AppStateListener"
  } );
  if ( isDebugMode( ) ) {
    logger.info( loadTime );
  }
  const { deviceStorageFull, showStorageFullAlert } = useDeviceStorageFull( );
  const [deviceStorageFullShown, setDeviceStorageFullShown] = useState( false );

  useEffect( ( ) => {
    const subscription = AppState.addEventListener( "change", nextAppState => {
      if ( nextAppState === "active" ) {
        // Check storage and show alert as needed when app becomes active
        if ( deviceStorageFull && !deviceStorageFullShown ) {
          showStorageFullAlert( );
          setDeviceStorageFullShown( true );
        }

        focusManager.setFocused( true );
      }
    } );

    return ( ) => subscription.remove( );
  }, [deviceStorageFull, deviceStorageFullShown, showStorageFullAlert] );

  return null;
};

export default AppStateListener;
