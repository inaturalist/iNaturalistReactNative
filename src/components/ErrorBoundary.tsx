import {
  Body1,
  Button,
  Heading1,
  Heading2
} from "components/SharedComponents";
import {
  SafeAreaView,
  ScrollView
} from "components/styledComponents";
import { t } from "i18next";
import React from "react";
import RNRestart from "react-native-restart";
import { log } from "sharedHelpers/logger";

const logger = log.extend( "ErrorBoundary" );

interface Props {
  children: React.ReactNode;
}

interface State {
  error: Error | null;
  info: React.ErrorInfo | null;
}

// https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary
class ErrorBoundary extends React.Component<Props, State> {
  constructor( props: Props ) {
    super( props );
    this.state = { error: null, info: null };
  }

  static getDerivedStateFromError( error: Error ) {
    // Update state so the next render will show the fallback UI.
    return { error };
  }

  componentDidCatch( error: Error, info: React.ErrorInfo ) {
    this.setState( { info } );
    if ( info?.componentStack ) {
      // componentStack is a little more informative than what's generally in
      // stack, so put that in there before logging
      error.stack = info.componentStack;
    }
    logger.error( error );
  }

  render() {
    if ( this.state.error ) {
      // TODO allow this to take a fallback component as a prop so individual
      // boundaries can customize their error states
      return (
        <SafeAreaView className="flex-1">
          <ScrollView className="p-5 bg-white">
            <Heading1 className="my-3">{ t( "Something-went-wrong" ) }</Heading1>
            <Body1>
              { t( "If-youre-seeing-this-error" ) }
            </Body1>
            <Button
              text={t( "RESTART-APP" )}
              onPress={( ) => RNRestart.restart()}
              className="my-5"
              level="focus"
            />
            <Heading2 className="mt-5">{ t( "Error" ) }</Heading2>
            <Body1>{ this.state.error.toString( ) }</Body1>
            <Body1>{ this.state.info?.componentStack }</Body1>
          </ScrollView>
        </SafeAreaView>
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
