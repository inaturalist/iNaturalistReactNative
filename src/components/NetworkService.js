import { log } from "sharedHelpers/logger";
import {
  useIconicTaxa,
  useObservationUpdatesWhenFocused, usePerformance
} from "sharedHooks";
import { isDebugMode } from "sharedHooks/useDebugMode";

import useTaxonCommonNames from "./hooks/useTaxonCommonNames";
import useWorkQueue from "./hooks/useWorkQueue";

const logger = log.extend( "NetworkService" );

const NetworkService = ( ) => {
  const { loadTime } = usePerformance( {
    screenName: "NetworkService"
  } );
  if ( isDebugMode( ) ) {
    logger.info( loadTime );
  }

  useObservationUpdatesWhenFocused( );

  useIconicTaxa( { reload: true } );
  // This only runs when App updates... which is rarely. It works for Settings
  // b/c it generally updates currentUser
  useWorkQueue( );
  useTaxonCommonNames( );

  return null;
};

export default NetworkService;
