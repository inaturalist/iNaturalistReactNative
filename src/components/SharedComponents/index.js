export { default as ActivityAnimation } from "./ActivityAnimation/ActivityAnimation";
export { default as ActivityCount } from "./ActivityCount/ActivityCount";
export { default as CommentsCount } from "./ActivityCount/CommentsCount";
export { default as IdentificationsCount } from "./ActivityCount/IdentificationsCount";
export { default as ActivityIndicator } from "./ActivityIndicator";
export { default as ButtonBar } from "./ButtonBar";
export { default as BackButton } from "./Buttons/BackButton";
export { default as Button } from "./Buttons/Button";
export { default as CloseButton } from "./Buttons/CloseButton";
export { default as EvidenceButton } from "./Buttons/EvidenceButton";
export { default as INatIconButton } from "./Buttons/INatIconButton";
export { default as RotatingINatIconButton } from "./Buttons/RotatingINatIconButton";
export { default as TransparentCircleButton } from "./Buttons/TransparentCircleButton";
export { default as Carousel } from "./Carousel";
export { default as CarouselDots } from "./CarouselDots";
export { default as Checkbox } from "./Checkbox";
export { default as CircleDots } from "./CircleDots";
export { default as ConfidenceInterval } from "./ConfidenceInterval";
export { default as DateDisplay } from "./DateDisplay";
export { default as DateTimePicker } from "./DateTimePicker";
export { default as DisplayTaxon } from "./DisplayTaxon";
export { default as DisplayTaxonName } from "./DisplayTaxonName";
export { default as Divider } from "./Divider/Divider";
export { default as CustomFlashList } from "./FlashList/CustomFlashList";
export { default as CustomRefreshControl } from "./FlashList/CustomRefreshControl";
export { default as FlashListEmptyWrapper } from "./FlashList/FlashListEmptyWrapper";
export { default as InfiniteScrollLoadingWheel } from "./FlashList/InfiniteScrollLoadingWheel";
export { default as FloatingActionBar } from "./FloatingActionBar";
export { default as HideView } from "./HideView";
export { default as IconicTaxonChooser } from "./IconicTaxonChooser";
export { default as IconicTaxonIcon } from "./IconicTaxonIcon";
export { default as INatIcon } from "./INatIcon";
export { default as InlineUser } from "./InlineUser/InlineUser";
export { default as InputField } from "./InputField";
export { default as KebabMenu } from "./KebabMenu";
export { default as KeyboardDismissableView } from "./KeyboardDismissableView";
export { default as LabelColonValue } from "./LabelColonValue";
export { default as DetailsMap } from "./Map/DetailsMap";
export { default as Map } from "./Map/Map";
export { default as MediaNavButtons } from "./MediaNavButtons";
export { default as Modal } from "./Modal";
export { default as Mortal } from "./Mortal";
export { default as ObservationLocation } from "./ObservationLocation";
export { default as ObsStatus } from "./ObsStatus";
export { default as OfflineNotice } from "./OfflineNotice";
export { default as OverlayHeader } from "./OverlayHeader";
export { default as OverviewCounts } from "./OverviewCounts";
export { default as PermissionGate } from "./PermissionGate";
export { default as PhotoCount } from "./PhotoCount";
export { default as QualityGradeStatus } from "./QualityGradeStatus/QualityGradeStatus";
export { default as RadioButtonRow } from "./RadioButtonRow";
export { default as ScrollableWithStickyHeader } from "./ScrollableWithStickyHeader";
export { default as ScrollViewWrapper } from "./ScrollViewWrapper";
export { default as SearchBar } from "./SearchBar";
export { default as BottomSheet } from "./Sheets/BottomSheet";
export { default as BottomSheetStandardBackdrop } from "./Sheets/BottomSheetStandardBackdrop";
export { default as PickerSheet } from "./Sheets/PickerSheet";
export { default as RadioButtonSheet } from "./Sheets/RadioButtonSheet";
export { default as TextInputSheet } from "./Sheets/TextInputSheet";
export { default as TextSheet } from "./Sheets/TextSheet";
export { default as WarningSheet } from "./Sheets/WarningSheet";
export { default as SimpleObservationLocation } from "./SimpleObservationLocation";
export { default as Tabs } from "./Tabs/Tabs";
export { default as TaxonResult } from "./TaxonResult";
export { default as TaxonSearch } from "./TaxonSearch";
export { default as Body1 } from "./Typography/Body1";
export { default as Body2 } from "./Typography/Body2";
export { default as Body3 } from "./Typography/Body3";
export { default as Body4 } from "./Typography/Body4";
export { default as Heading1 } from "./Typography/Heading1";
export { default as Heading2 } from "./Typography/Heading2";
export { default as Heading3 } from "./Typography/Heading3";
export { default as Heading4 } from "./Typography/Heading4";
export { default as Heading5 } from "./Typography/Heading5";
export { default as Heading6 } from "./Typography/Heading6";
export { default as List2 } from "./Typography/List2";
export { default as P } from "./Typography/P";
export { default as Subheading1 } from "./Typography/Subheading1";
export { default as Subheading2 } from "./Typography/Subheading2";
export { default as UnderlinedLink } from "./Typography/UnderlinedLink";
export { default as UploadProgressBar } from "./UploadProgressBar";
export { default as UploadStatus } from "./UploadStatus/UploadStatus";
export { default as UserIcon } from "./UserIcon";
export { default as UserText } from "./UserText";
export { default as ViewWrapper } from "./ViewWrapper";
