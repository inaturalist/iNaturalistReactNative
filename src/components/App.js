// @flow

import { useNavigation } from "@react-navigation/native";
import RootDrawerNavigator from "navigation/rootDrawerNavigator";
import type { Node } from "react";
import React, { useCallback } from "react";
import { log } from "sharedHelpers/logger";
import {
  useCurrentUser,
  usePerformance,
  useShare
} from "sharedHooks";
import { isDebugMode } from "sharedHooks/useDebugMode";

import AppStateListener from "./AppStateListener";
import useLinking from "./hooks/useLinking";
import NetworkService from "./NetworkService";
import StartupService from "./StartupService";

const logger = log.extend( "App" );

type SharedItem = {
  mimeType: string,
  data: string | Array<string>
};

const handleShare = ( navigation, item: ?SharedItem ) => {
  if ( !item ) {
    // user hasn't shared any items
    return;
  }

  const { mimeType, data } = item;

  if ( !mimeType && !data ) {
    // user hasn't shared any images
    return;
  }

  // show user a loading animation screen (like PhotoLibrary)
  // while observations are created
  navigation?.navigate( "PhotoSharing", { item } );
};

type Props = {
  // $FlowIgnore
  children?: unknown,
};

// this children prop is here for the sake of testing with jest
// normally we would never do this in code
const App = ( { children }: Props ): Node => {
  const navigation = useNavigation( );
  const { loadTime } = usePerformance( {
    screenName: "App"
  } );
  if ( isDebugMode( ) ) {
    logger.info( loadTime );
  }

  // attempting to make sure that navigation is only called once
  // for performance reasons
  const onShare = useCallback(
    item => handleShare( navigation, item ),
    [navigation]
  );

  const currentUser = useCurrentUser( );

  useLinking( currentUser );
  useShare( onShare );

  // this children prop is here for the sake of testing with jest
  // normally we would never do this in code
  return (
    <>
      <StartupService />
      <NetworkService />
      <AppStateListener />
      {children || <RootDrawerNavigator />}
    </>
  );
};

export default App;
