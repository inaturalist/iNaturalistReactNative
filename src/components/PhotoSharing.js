// @flow

import { useNavigation, useRoute } from "@react-navigation/native";
import { ActivityAnimation, ViewWrapper } from "components/SharedComponents";
import { View } from "components/styledComponents";
import type { Node } from "react";
import React, { useCallback, useEffect, useState } from "react";
import { Alert, Platform } from "react-native";
import Observation from "realmModels/Observation";
import useStore from "stores/useStore";

const PhotoSharing = ( ): Node => {
  const navigation = useNavigation( );
  const { params } = useRoute( );
  const { item } = params;
  const sharedText = item.extraData?.userInput;
  const resetObservationFlowSlice = useStore( state => state.resetObservationFlowSlice );
  const prepareObsEdit = useStore( state => state.prepareObsEdit );
  const setPhotoImporterState = useStore( state => state.setPhotoImporterState );
  const isAdvancedSuggestionsMode = useStore( state => state.layout.isAdvancedSuggestionsMode );
  const [navigationHandled, setNavigationHandled] = useState( null );

  const createObservationAndNavToObsEdit = useCallback( async photoUris => {
    try {
      const newObservation = await Observation.createObservationWithPhotos( photoUris );
      newObservation.description = sharedText;
      prepareObsEdit( newObservation );
      if ( isAdvancedSuggestionsMode ) {
        navigation.navigate(
          "NoBottomTabStackNavigator",
          { screen: "Suggestions", params: { lastScreen: "PhotoSharing" } }
        );
      } else {
        navigation.navigate( "NoBottomTabStackNavigator", { screen: "ObsEdit" } );
      }
    } catch ( e ) {
      Alert.alert(
        "Photo sharing failed: couldn't create new observation:",
        e
      );
    }
  }, [
    navigation,
    prepareObsEdit,
    sharedText,
    isAdvancedSuggestionsMode
  ] );

  useEffect( ( ) => {
    const { mimeType, data } = item;

    if ( Platform.OS === "android" && !mimeType.startsWith( "image/" ) ) {
      Alert.alert( "Android photo share failed: not an image file" );
      return;
    }

    // Move to ObsEdit screen (new observation, with shared photos).

    // Create a new observation with multiple shared photos (one or more)
    let photoUris;

    // data is returned as a string for a single photo on Android
    // and an object with an array of data strings on iOS, i.e.
    // [{"data": "", "", "mimeType": "image/jpeg]
    if ( Array.isArray( data ) ) {
      photoUris = data;
    } else {
      photoUris = [data];
    }

    if ( Platform.OS === "android" ) {
      photoUris = photoUris.map( x => ( { image: { uri: x } } ) );
    } else {
      photoUris = photoUris
        .filter( x => x.mimeType && x.mimeType.startsWith( "image/" ) )
        .map( x => ( { image: { uri: x.data } } ) );
    }

    if ( photoUris.length === 1 ) {
      // Only one photo - go to ObsEdit directly
      createObservationAndNavToObsEdit( photoUris );
    } else {
      // Go to GroupPhotos screen
      const firstObservationDefaults = { description: sharedText };
      setPhotoImporterState( {
        photoLibraryUris: photoUris.map( x => x.image.uri ),
        groupedPhotos: photoUris.map( photo => ( {
          photos: [photo]
        } ) ),
        firstObservationDefaults
      } );
      navigation.navigate( "NoBottomTabStackNavigator", { screen: "GroupPhotos" } );
    }
  }, [
    createObservationAndNavToObsEdit,
    item,
    navigation,
    resetObservationFlowSlice,
    setPhotoImporterState,
    sharedText
  ] );

  // When the user leaves this screen, we record the fact that navigation was handled...
  useEffect( ( ) => {
    const unsubscribe = navigation.addListener( "blur", ( ) => {
      setNavigationHandled( true );
    } );
    return unsubscribe;
  }, [navigation] );
  // ...and if they focus on this screen again, that means they backed out of
  // obs edit and need to back to the previous screen in the nav
  useEffect( ( ) => {
    const unsubscribe = navigation.addListener( "focus", ( ) => {
      if ( navigationHandled ) navigation.goBack( );
    } );
    return unsubscribe;
  }, [
    navigation,
    navigationHandled
  ] );

  return (
    <ViewWrapper testID="PhotoSharing">
      <View className="flex-1 w-full h-full justify-center items-center">
        <ActivityAnimation />
      </View>
    </ViewWrapper>
  );
};

export default PhotoSharing;
