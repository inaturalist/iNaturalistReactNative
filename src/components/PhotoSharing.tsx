import { CommonActions, useNavigation, useRoute } from "@react-navigation/native";
import { ActivityAnimation, ViewWrapper } from "components/SharedComponents";
import { View } from "components/styledComponents";
import React, {
  useCallback, useEffect, useRef, useState
} from "react";
import { Alert, Platform } from "react-native";
import Observation from "realmModels/Observation";
import { useLayoutPrefs } from "sharedHooks";
import useStore from "stores/useStore";

const PhotoSharing = ( ) => {
  const previousItem = useRef( null );
  const navigation = useNavigation( );
  const { params } = useRoute( );
  const { item } = params;
  const sharedText = item.extraData?.userInput;
  const resetObservationFlowSlice = useStore( state => state.resetObservationFlowSlice );
  const prepareObsEdit = useStore( state => state.prepareObsEdit );
  const setPhotoImporterState = useStore( state => state.setPhotoImporterState );
  const { screenAfterPhotoEvidence, isDefaultMode } = useLayoutPrefs();
  const [navigationHandled, setNavigationHandled] = useState( null );

  const resetNavigator = useCallback( screen => navigation.dispatch(
    CommonActions.reset( {
      index: 0,
      routes: [
        {
          name: "NoBottomTabStackNavigator",
          state: {
            index: 0,
            routes: [
              {
                name: screen,
                params: { lastScreen: "PhotoSharing" }
              }
            ]
          }
        }
      ]
    } )
  ), [navigation] );

  const createObservationAndNavigate = useCallback( async photoUris => {
    try {
      const newObservation = await Observation.createObservationWithPhotos( photoUris );
      newObservation.description = sharedText;
      prepareObsEdit( newObservation );

      return resetNavigator( isDefaultMode
        ? "Match"
        : screenAfterPhotoEvidence );
    } catch ( e ) {
      Alert.alert(
        "Photo sharing failed: couldn't create new observation:",
        e
      );
      return null;
    }
  }, [sharedText, prepareObsEdit, isDefaultMode, resetNavigator, screenAfterPhotoEvidence] );

  useEffect( ( ) => {
    const { mimeType, data } = item;

    if ( Platform.OS === "android" && !mimeType.startsWith( "image/" ) ) {
      Alert.alert( "Android photo share failed: not an image file" );
      return;
    }

    // when sharing, we need to reset zustand like we do while
    // navigating through the AddObsModal
    resetObservationFlowSlice( );

    // Create a new observation with multiple shared photos (one or more)
    let photoUris;

    // data is returned as a string for a single photo on Android
    // and an object with an array of data strings on iOS, i.e.
    // [{"data": "", "", "mimeType": "image/jpeg]
    if ( Array.isArray( data ) ) {
      photoUris = data;
    } else {
      photoUris = [data];
    }

    if ( Platform.OS === "android" ) {
      photoUris = photoUris.map( x => ( { image: { uri: x } } ) );
    } else {
      photoUris = photoUris
        .filter( x => x.mimeType && x.mimeType.startsWith( "image/" ) )
        .map( x => ( { image: { uri: x.data } } ) );
    }

    if ( photoUris.length === 1 ) {
      createObservationAndNavigate( photoUris );
    } else {
      // Go to GroupPhotos screen
      const firstObservationDefaults = { description: sharedText };
      setPhotoImporterState( {
        photoLibraryUris: photoUris.map( x => x.image.uri ),
        groupedPhotos: photoUris.map( photo => ( {
          photos: [photo]
        } ) ),
        firstObservationDefaults
      } );
      resetNavigator( "GroupPhotos" );
    }
  }, [
    createObservationAndNavigate,
    item,
    navigation,
    resetObservationFlowSlice,
    setPhotoImporterState,
    sharedText,
    resetNavigator
  ] );

  // When the user leaves this screen, we record the fact that navigation was handled...
  useEffect( ( ) => {
    const unsubscribe = navigation.addListener( "blur", ( ) => {
      setNavigationHandled( true );
    } );
    return unsubscribe;
  }, [navigation] );

  // ...and if they focus on this screen again, that means they backed out of
  // obs edit and need to back to the previous screen in the nav
  useEffect( () => {
    const unsubscribe = navigation.addListener( "focus", () => {
      const isNewShare = item !== previousItem.current;

      // don't navigate backwards if there are new items shared, otherwise
      // we end up back on the home screen without the user ever seeing their shared items
      if ( isNewShare ) {
        setNavigationHandled( false );
        previousItem.current = item;
        return;
      }

      if ( navigationHandled ) {
        navigation.goBack( );
      }
    } );
    return unsubscribe;
  }, [navigation, navigationHandled, item] );

  return (
    <ViewWrapper testID="PhotoSharing">
      <View className="flex-1 w-full h-full justify-center items-center">
        <ActivityAnimation />
      </View>
    </ViewWrapper>
  );
};

export default PhotoSharing;
