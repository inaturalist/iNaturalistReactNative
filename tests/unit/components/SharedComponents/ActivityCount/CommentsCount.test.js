import { render, screen } from "@testing-library/react-native";
import { CommentsCount } from "components/SharedComponents";
import React from "react";

const count = 1;

describe( "CommentsCount", () => {
  it( "renders default reliably", () => {
    // Snapshot test
    render( <CommentsCount count={count} /> );

    expect( screen ).toMatchSnapshot();
  } );

  it( "renders white reliably", () => {
    // Snapshot test
    render( <CommentsCount count={count} white /> );

    expect( screen ).toMatchSnapshot();
  } );

  it( "renders filled reliably", () => {
    // Snapshot test
    render( <CommentsCount count={count} filled /> );

    expect( screen ).toMatchSnapshot();
  } );

  // a11y test
  it( "should not have accessibility errors", () => {
    // const activityCount = <CommentsCount count={count} />;

    // Disabled during the update to RN 0.78
    // expect( activityCount ).toBeAccessible();
  } );
} );
