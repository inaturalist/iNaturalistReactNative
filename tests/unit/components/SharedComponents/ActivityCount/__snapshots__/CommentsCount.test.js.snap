// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentsCount renders default reliably 1`] = `
<View
  accessibilityLabel="1 comment"
  accessible={true}
  style={
    [
      [
        {
          "flexDirection": "row",
        },
        {
          "alignItems": "center",
        },
      ],
    ]
  }
>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      [
        {
          "color": "#454545",
          "fontSize": 14,
        },
        null,
        {
          "fontFamily": "INatIcon",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        {},
      ]
    }
  >
    
  </Text>
  <Text
    maxFontSizeMultiplier={2}
    style={
      [
        {
          "color": "#454545",
        },
        {
          "fontFamily": "Lato-Regular",
        },
        [
          {
            "textAlign": "left",
          },
          [
            {
              "fontSize": 13,
              "lineHeight": 18,
            },
            {
              "fontFamily": "Lato-Medium",
            },
            [
              {
                "marginLeft": 6,
              },
            ],
          ],
        ],
      ]
    }
  >
    1
  </Text>
</View>
`;

exports[`CommentsCount renders filled reliably 1`] = `
<View
  accessibilityLabel="1 comment"
  accessible={true}
  style={
    [
      [
        {
          "flexDirection": "row",
        },
        {
          "alignItems": "center",
        },
      ],
    ]
  }
>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      [
        {
          "color": "#454545",
          "fontSize": 14,
        },
        null,
        {
          "fontFamily": "INatIcon",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        {},
      ]
    }
  >
    
  </Text>
  <Text
    maxFontSizeMultiplier={2}
    style={
      [
        {
          "color": "#454545",
        },
        {
          "fontFamily": "Lato-Regular",
        },
        [
          {
            "textAlign": "left",
          },
          [
            {
              "fontSize": 13,
              "lineHeight": 18,
            },
            {
              "fontFamily": "Lato-Medium",
            },
            [
              {
                "marginLeft": 6,
              },
            ],
          ],
        ],
      ]
    }
  >
    1
  </Text>
</View>
`;

exports[`CommentsCount renders white reliably 1`] = `
<View
  accessibilityLabel="1 comment"
  accessible={true}
  style={
    [
      [
        {
          "flexDirection": "row",
        },
        {
          "alignItems": "center",
        },
      ],
    ]
  }
>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      [
        {
          "color": "#ffffff",
          "fontSize": 14,
        },
        null,
        {
          "fontFamily": "INatIcon",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        {},
      ]
    }
  >
    
  </Text>
  <Text
    maxFontSizeMultiplier={2}
    style={
      [
        {
          "color": "#454545",
        },
        {
          "fontFamily": "Lato-Regular",
        },
        [
          {
            "textAlign": "left",
          },
          [
            {
              "fontSize": 13,
              "lineHeight": 18,
            },
            {
              "fontFamily": "Lato-Medium",
            },
            [
              {
                "marginLeft": 6,
              },
              {
                "color": "#ffffff",
              },
            ],
          ],
        ],
      ]
    }
  >
    1
  </Text>
</View>
`;
